### PR TITLE
PartTemplate UI and Backend changes

### PR DESCRIPTION
**Pull request by RossStewart**
Commit for /inventory/part-templates/

### CODE REVIEW

File: src/Facade/ResourceBuilders/PartTemplateResourceBuilder.cs [hunk 1]
```diff
@@ -27,7 +29,9 @@ public PartTemplateResource Build(PartTemplate model)
                            BomType = model.BomType,
                            AssemblyTechnology = model.AssemblyTechnology,
                            AllowPartCreation = model.AllowPartCreation,
-                           ParetoCode = model.ParetoCode
+                           ParetoCode = model.ParetoCode,
+                           Links = this.BuildLinks(model).ToArray()
+
```
Comment (lewisrenfrew) [inline src/Facade/ResourceBuilders/PartTemplateResourceBuilder.cs:34]: blank line

File: src/Service.Host/client/src/components/parts/PartTemplate.js [hunk 2]
```diff
@@ -0,0 +1,534 @@
...
+    setEditStatus,
+    setSnackbarVisible,
+    inDialogBox,
+    privileges,
+    closeDialog,
+    productAnalysisCodeSearchResults,
+    productAnalysisCodesSearchLoading,
+    searchProductAnalysisCodes,
+    clearProductAnalysisCodesSearch,
+    assemblyTechnology
```
Comment (lewisrenfrew) [inline src/Service.Host/client/src/components/parts/PartTemplate.js:36]: would rename this assemblyTechnologies for clarity

File: src/Service.Host/client/src/components/parts/PartTemplate.js [hunk 3]
```diff
@@ -0,0 +1,534 @@
...
+            display: 'inline-block',
+            width: '2em'
+        },
+        marginTopWiderLinkButton: {
+            marginTop: '-14px',
+            display: 'inline-block'
+        },
+        displayInline: {
+            display: 'inline'
+        },
```
Comment (lewisrenfrew) [inline src/Service.Host/client/src/components/parts/PartTemplate.js:104]: I don't think any of these classes other than thinPage were used

File: src/Service.Host/client/src/components/parts/PartTemplate.js [hunk 4]
```diff
@@ -0,0 +1,534 @@
...
+                            onClose={() => setSnackbarVisible(false)}
+                            message="Save Successful"
+                        />
+
+                        <Grid item xs={4}>
+                            <InputField
+                                fullWidth
+                                value={partTemplate.partRoot}
+                                label="Part Root"
+                                maxLength={20}
```
Comment (lewisrenfrew) [inline src/Service.Host/client/src/components/parts/PartTemplate.js:189]: is this maxLength correct (check the db table)

File: src/Service.Host/client/src/components/parts/PartTemplate.js [hunk 5]
```diff
@@ -0,0 +1,534 @@
...
+                                items={[
+                                    { id: 'A', displayText: 'A - BOUGHT IN, TOP 80% OF SPEND' },
+                                    { id: 'B', displayText: 'B - BOUGHT IN, NEXT 15% OF SPEND' },
+                                    { id: 'C', displayText: 'C - BOUGHT IN, LEAST 5% OF SPEND' },
+                                    { id: 'D', displayText: 'D - BOUGHT IN. NO USAGE' },
+                                    { id: 'E', displayText: 'E - LINN PRODUCED PARTS' },
+                                    { id: 'L', displayText: 'L - LOEWE TVS AND ACCESSORIES' },
+                                    { id: 'R', displayText: 'R - ALL RECORDS, TAPES AND CDS' },
+                                    { id: 'U', displayText: 'U - UNSOURCED COMPONENTS' },
+                                    { id: 'X', displayText: 'X - PARTS OUTWITH PARETO CODES' }
```
Comment (lewisrenfrew) [inline src/Service.Host/client/src/components/parts/PartTemplate.js:429]: This one probably could do with a real database lookup for the results. We spoke about doing that in your next PR already though so feel free to merge as is

File: src/Service.Host/client/src/components/parts/PartTemplate.js [hunk 6]
```diff
@@ -0,0 +1,534 @@
...
+                        </Grid>
+                    </>
+                )
+            )}
+        </Grid>
+    );
+
+    return (
+        <>
+            {inDialogBox ? (
```
Comment (lewisrenfrew) [inline src/Service.Host/client/src/components/parts/PartTemplate.js:457]: same as above

File: src/Service.Host/client/src/components/parts/PartTemplate.js [hunk 7]
```diff
@@ -0,0 +1,534 @@
...
+    closeDialog: PropTypes.func,
+    productAnalysisCodeSearchResults: PropTypes.arrayOf(productAnalysisCodeShape),
+    searchProductAnalysisCodes: PropTypes.func.isRequired,
+    clearProductAnalysisCodesSearch: PropTypes.func,
+    productAnalysisCodesSearchLoading: PropTypes.bool,
+    assemblyTechnology: PropTypes.arrayOf(PropTypes.shape({}))
+};
+
+PartTemplate.defaultProps = {
+    item: {
```
Comment (lewisrenfrew) [inline src/Service.Host/client/src/components/parts/PartTemplate.js:503]: would just default this to null

File: src/Service.Host/client/src/components/parts/PartTemplate.js [hunk 8]
```diff
@@ -0,0 +1,534 @@
...
+    snackbarVisible: false,
+    loading: false,
+    itemError: null,
+    itemId: null,
+    inDialogBox: false,
+    closeDialog: null,
+    productAnalysisCodeSearchResults: [],
+    productAnalysisCodesSearchLoading: false,
+    assemblyTechnology: [],
+    clearProductAnalysisCodesSearch: () => {}
```
Comment (lewisrenfrew) [inline src/Service.Host/client/src/components/parts/PartTemplate.js:531]: I'd make this a required prop and scrub this default

File: src/Service.Host/client/src/components/parts/PartTemplatesSearch.js [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+            <Grid container spacing={3}>
+                <Grid item xs={7}>
+                    <Typography variant="h3">Part Template Utility</Typography>
+                </Grid>
+                <Grid item xs={1}>
+                    <LinkButton
+                        text="Create"
+                        to="/inventory/part-templates/create"
+                        disabled={!canCreate()}
+                        tooltip={canCreate() ? null : 'You are not authorised to create parts.'}
```
Comment (lewisrenfrew) [inline src/Service.Host/client/src/components/parts/PartTemplatesSearch.js:32]: should be 'You are not authorised to create parts.', technically.

File: src/Service.Host/client/src/containers/parts/PartTemplate.js [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+const mapStateToProps = (state, { match }) => ({
+    item: creating(match) ? null : partTemplateSelectors.getItem(state),
+    itemId: creating(match) ? null : match.params.id,
+    editStatus: creating(match) ? 'create' : partTemplateSelectors.getEditStatus(state),
+    loading: partTemplateSelectors.getLoading(state),
+    snackbarVisible: partTemplateSelectors.getSnackbarVisible(state),
+    itemError: getItemError(state, itemTypes.partTemplate.item),
+    privileges: getPrivileges(state),
+    userNumber: getUserNumber(state),
+    previousPaths: getPreviousPaths(state),
```
Comment (lewisrenfrew) [inline src/Service.Host/client/src/containers/parts/PartTemplate.js:29]: did we use this previousPaths prop?

File: src/Service.Host/client/src/containers/parts/PartTemplate.js [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+    productAnalysisCodeSearchResults: productAnalysisCodesSelectors
+        .getSearchItems(state)
+        .map(c => ({ name: c.productCode, description: c.description })),
+    productAnalysisCodesSearchLoading: productAnalysisCodesSelectors.getSearchLoading(state),
+    assemblyTechnology: assemblyTechnologiesSelectors.getItems(state)
```
Comment (lewisrenfrew) [inline src/Service.Host/client/src/containers/parts/PartTemplate.js:34]: would rename to assemblyTechnologies (to match component)

File: src/Service.Host/client/src/containers/parts/PartTemplates.js [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+import partTemplatesActions from '../../actions/partTemplatesActions';
+import partTemplatesSelectors from '../../selectors/partTemplatesSelectors';
+import partTemplateSelectors from '../../selectors/partTemplateSelectors';
+import PartTemplatesSearch from '../../components/parts/PartTemplatesSearch';
+import { getPrivileges } from '../../selectors/userSelectors';
+
+const mapStateToProps = (state, { match }) => ({
+    items: partTemplatesSelectors.getItems(state),
+    loading: partTemplatesSelectors.getLoading(state),
+    linkToSources: match?.url?.endsWith('/sources'),
```
Comment (lewisrenfrew) [inline src/Service.Host/client/src/containers/parts/PartTemplates.js:12]: is linkToSources doing anything?

File: src/Service.Host/client/src/containers/parts/PartTemplates.js [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+    partTemplate: partTemplateSelectors.getItem(state)
+});
+
+const initialise = () => dispatch => {
+    dispatch(partTemplatesActions.fetch());
+};
+
+const mapDispatchToProps = {
+    initialise,
+    fetchItems: partTemplatesActions.fetchByQueryString,
```
Comment (lewisrenfrew) [inline src/Service.Host/client/src/containers/parts/PartTemplates.js:24]: was fetchItems used? Seems  you already fetched items in the initialise() so no need to pass it to the component as a prop I'd imagine

**Review by lewisrenfrew — APPROVED** [review]
Loosk good, suggested some changes you may or may not want to make :)

File: src/Service.Host/client/src/components/parts/PartTemplatesSearch.js [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+
+function PartTemplateSearch({ privileges, partTemplates }) {
+    const [template, setTemplate] = useState();
+
+    const createUrl = () => {
+        return `/inventory/part-templates/${template}`;
+    };
+    const canCreate = () => {
+        if (!(privileges.length < 1)) {
+            return privileges.some(priv => priv === 'part.admin');
```
Comment (lewisrenfrew) [inline src/Service.Host/client/src/components/parts/PartTemplatesSearch.js:16]: I'd stick a link on the Resource to /create if the user has the correct permissions and then use the presence or absence of that link to decide canCreate(). This works too obviously but we're trying to move more towards the HATEOAS approach as Richard said

**Comment by RossStewart** [comment]
> Loosk good, suggested some changes you may or may not want to make :)

will sort these :)

File: src/Service.Host/client/src/components/parts/PartTemplate.js [hunk 15]
```diff
@@ -0,0 +1,534 @@
...
+        setPartTemplate(x => ({ ...x, [propertyName]: newValue }));
+    };
+
+    const handleProductCodeChange = product => {
+        handleFieldChange('productCode', product.name);
+        handleFieldChange('productAnalysisCodeDescription', product.description);
+    };
+
+    const allowedToEdit = () => {
+        return privileges?.some(priv => priv === 'part.admin');
```
Comment (lewisrenfrew) [inline src/Service.Host/client/src/components/parts/PartTemplate.js:154]: same as for creating, I'd stick a rel="edit" link on the resource in the builders BuildLinks and then use the presence or absence of that link to decide allowedToEdit()... mostly to keep Richard happy

**Pull request closed without merging**